csharp/src/Ice/SlicBinaryConnection.cs

```diff
@@ -107,164 +161,233 @@ public async ValueTask CloseAsync(Exception exception, CancellationToken cancel)
 
         public async ValueTask<(long StreamId, IncomingFrame? Frame, bool Fin)> ReceiveAsync(CancellationToken cancel)
         {
+            FrameType type;
+            ArraySegment<byte> data;
             while (true)
             {
-                ArraySegment<byte> buffer = await PerformReceiveFrameAsync().ConfigureAwait(false);
-                (int requestId, IncomingFrame? frame) = ParseFrame(buffer);
-                if (frame != null)
+                // Read Slic frame header
+                (type, data) = await ReceiveFrameAsync(CancellationToken.None).ConfigureAwait(false);
+                switch (type)
                 {
-                    return (StreamId: requestId, Frame: frame, Fin: requestId == 0 || frame is IncomingResponseFrame);
+                    case FrameType.Ping:
+                    {
+                        ValueTask task = PrepareAndSendFrameAsync(FrameType.Pong, null, CancellationToken.None);
+                        _heartbeatCallback();
+                        break;
+                    }
+                    case FrameType.Pong:
+                    {
+                        // TODO: setup a timer to expect pong frame response?
+                        break;
+                    }
+                    case FrameType.Stream:
+                    case FrameType.StreamLast:
+                    {
+                        (long streamId, int streamIdSize) = data.AsReadOnlySpan().ReadVarLong();
+                        data = data.Slice(streamIdSize);
+                        IncomingFrame frame = ParseIce2Frame(data);
+                        ProtocolTrace.TraceFrame(Endpoint.Communicator, streamId, data.Count, frame);
+                        return (StreamId: streamId, Frame: frame, Fin: type == FrameType.StreamLast);
+                    }
+                    case FrameType.ResetStream:
+                    {
+                        var istr = new InputStream(data, Encoding);
+                        long streamId = istr.ReadVarLong();
+                        long reason = istr.ReadVarLong();
+                        // TODO: release resources allocated for the stream.
+                        break;
+                    }
+                    case FrameType.Close:
+                    {
+                        var istr = new InputStream(data, Encoding);
+                        long code = istr.ReadVarLong();
+                        string reason = istr.ReadString();
+                        throw new ConnectionClosedByPeerException();
+                    }
+                    default:
+                    {
+                        throw new InvalidDataException($"unexpected Slic frame with frame type `{type}'");
+                    }
                 }
             }
         }
 
         public long NewStream(bool bidirectional) => bidirectional ? ++_nextStreamId : 0;
```

Comment:
Should we use the same stream IDs as Quic? In particular, not 0 for oneways.

Reply:
Yes, definitely! For now, we're just doing the same as Ice1 but this will be changed as soon as we start keeping track of the active streams.